### PR TITLE
Improves Sort trailing numbers in tag names numerically ...

### PR DESCRIPTION
**Pull request by HelioGuilherme66**
in View All Tags #1579

Proposed by @alexisbietti.
(code is PEP8 compliant)

### CODE REVIEW

**Comment by HelioGuilherme66** [comment]
Please hold merging this one. I will incorporate PR #1593 because they touch same file.

**Comment by HelioGuilherme66** [comment]
Please merge. Thanks.

**Comment by yanne** [comment]
This branch still contains 11 commits, most of which are noise. The commits should be squashed with `git rebase -i` possibly first rebasing on current master by doing `git fetch origin-ride && git rebase origin-ride/master` before this can be merged.

**Comment by HelioGuilherme66** [comment]
I believe this is it! Merge me!

File: src/robotide/ui/tagdialogs.py [hunk 1]
```diff
@@ -121,10 +130,10 @@ def _execute(self):
...
 
     def update_footer(self):
-        footer_string = "Total tests %d, Tests with tags %d, Unique tags %d," \
-                        " Currently selected tests %d" % \
-            (self.total_test_cases, len(self.tagged_test_cases),
-             self.unique_tags, len(self.selected_tests))
+        footer_string = "Total tests %d, Tests with tags %d, Unique tags %d, \
+Currently selected tests %d" % (self.total_test_cases,
+                                len(self.tagged_test_cases), self.unique_tags,
+                                len(self.selected_tests))
```
Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: Is this just a formatting change? Should preferably be in a separate commit (or even pr). More importantly,  I don't think cutting a string like

```python
    "Example string \
here is."
```

is a good idea. I'd use either

```python
    "Example string " \
    "here is."
```

or probably

```python
    ("Example string "
     "here is.")
```

The latter is also what [PEP-8](https://www.python.org/dev/peps/pep-0008/#maximum-line-length) recommends:

> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

File: src/robotide/ui/tagdialogs.py [hunk 2]
```diff
@@ -304,15 +312,14 @@ def add_tag(self, tag_to_tests):
...
-        insert statements.
+        the model does not. The entry in the model is referenced by a new id
+        generated by wx.NewId(). This prevents overwriting entries when
+        combining delete and insert statements.
 
         Args:
             index: An int marking the position to insert the row at.
             tag_to_tests: A tuple mapping tests(list, index 0) to a
-            tag(str, index 1).
+                          tag(str, index 1).
```
Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: What happened here?

File: src/robotide/ui/tagdialogs.py [hunk 3]
```diff
@@ -161,8 +170,7 @@ def _search_for_tags(self):
         isreversed = (self.sort_state[1] != 1)
         self.total_test_cases = len(self._test_cases)
         self._results = sorted(self._unique_tags.items(),
-                               key=lambda x: str(x[0]).lower,
-                               reverse=isreversed)
+                               key=lambda x: str(x[0]).lower, reverse=isreversed)
```
Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: 1. I don't think this kind of changes should be done in general.

2. I actually like the original formatting better.

3. That `str(x[0])` call looks dangerous. It will fail if any `x[0]` contains non-ASCII characters. Are all these items always Unicode or byte strings? If yes, just using `x[0].lower` would work.

4. `item` would be a better name than `x` when sorting `.items()`.

File: src/robotide/ui/tagdialogs.py (hunk 2 above)
Comment (HelioGuilherme66) [inline src/robotide/ui/tagdialogs.py]: It is just indentation for the doc.

File: src/robotide/ui/tagdialogs.py [hunk 4]
```diff
@@ -100,6 +101,14 @@ def _make_bindings(self):
         self._tags_list.Bind(wx.EVT_LIST_ITEM_RIGHT_CLICK, self.OnRightClick)
         self._tags_list.Bind(wx.EVT_LIST_COL_CLICK, self.OnColClick)
 
+    def _tag_name_for_sort(self, tag_name):
+        # Make tag_name lowercase for the sorting algorithm only
+        tag_name = tag_name.lower()
+        # Compare trailing numbers as numeric values
+        m = re.search(r'([^0-9]*)([0-9]*)', tag_name)
```
Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: According to the comment this regexp tries to to find trailing numbers, but the regexp itself actually finds *any* number. Is comment or code correct??

File: src/robotide/ui/tagdialogs.py (hunk 3 above)
Comment (HelioGuilherme66) [inline src/robotide/ui/tagdialogs.py]: 1 and 2. It was an unnoticed change when squashing, I also like better as it was.

File: src/robotide/ui/tagdialogs.py (hunk 2 above)
Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: Oh, I see. It was indented to match `: ` on the above line. I guess it makes sense. Having changes like this in same commits as actual features is nevertheless confusing.

File: src/robotide/ui/tagdialogs.py (hunk 3 above)
Comment (HelioGuilherme66) [inline src/robotide/ui/tagdialogs.py]: 3. Was indeed a bad piece of code. I have also updated the test file to have unicode chars.
4. Oops I forgot to change the var name (coming soon another forced commit!)

File: src/robotide/ui/tagdialogs.py [hunk 5]
```diff
@@ -100,6 +101,14 @@ def _make_bindings(self):
         self._tags_list.Bind(wx.EVT_LIST_ITEM_RIGHT_CLICK, self.OnRightClick)
         self._tags_list.Bind(wx.EVT_LIST_COL_CLICK, self.OnColClick)
 
+    def _tag_name_for_sort(self, tag_name):
+        # Make tag_name lowercase for the sorting algorithm only
+        tag_name = tag_name.lower()
+        # Compare numbers as numeric values
+        m = re.search(r'([^0-9]*)([0-9]*)', tag_name)
+        g = m.groups()
+        return g[0], int(g[1] or 0)
```
Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: This doesn't seem to handle tags like `1foo` and `1bar` or `foo1x` and `foo1y` correctly.

**Comment by pekkaklarck** [comment]
Not really related to this PR, but looking at the code I noticed some weird code at https://github.com/robotframework/RIDE/blob/master/src/robotide/ui/tagdialogs.py#L144:

```python
        self._unique_tags = utils.NormalizedDict()
        self._tagit = utils.NormalizedDict()
        self._test_cases = list()
```

The weirdness is that none of these seem to be used outside the method where they are created and `self._tagit` is ever only populated. Do you guys see any reason why not to change the above to:

```python
        unique_tags = utils.NormalizedDict()
        test_cases = []
```

If that is a good change, could you @HelioGuilherme66 commit that along with the clean-ups you have spotted separately?

**Comment by HelioGuilherme66** [comment]
Done the changes, except at self._tagit because it is used in Rename and Delete code.

**Comment by yanne** [comment]
The `self._tagit` initialisation is definitely poor style, I'd prefer all instance variables to be declared in the `__init__` method. However, they are used and can't be just removed.

**Comment by pekkaklarck** [comment]
Yeah, my bad. I wasn't looking for the whole file and didn't see other usages. Attributes should definitely be defined already in `__init__`, though.

`tagit` name is also pretty weird. It actually sounds like Finnish and should probably be renamed to `tags`.

File: src/robotide/ui/tagdialogs.py (hunk 5 above)
Comment (HelioGuilherme66) [inline src/robotide/ui/tagdialogs.py]: The following images show the differences:
Before:
tag_sort_old_algorithm
After:
tag_sort_new_algorithm

Is it acceptable?

Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: No, I don't think implementation that ignores everything after the first group of digits is acceptable.

Comment (HelioGuilherme66) [inline src/robotide/ui/tagdialogs.py]: Changed code to `m = re.search(r"(\w*)([0-9]*)", tag_name)` and now looks better (keeping the trailing numeric order):
tag_sort_mod_algorithm

**Comment by HelioGuilherme66** [comment]
Final changes I hope ;)

File: src/robotide/ui/tagdialogs.py [hunk 6]
```diff
@@ -100,6 +102,15 @@ def _make_bindings(self):
         self._tags_list.Bind(wx.EVT_LIST_ITEM_RIGHT_CLICK, self.OnRightClick)
         self._tags_list.Bind(wx.EVT_LIST_COL_CLICK, self.OnColClick)
 
+    def _tag_name_for_sort(self, tag_name):
+        # Make tag_name lowercase for the sorting algorithm only
+        tag_name = tag_name.lower()
+        # Compare numbers as numeric values
+        # m = re.search(r"([^0-9]*)([0-9]*)", tag_name)
+        m = re.search(r"(\w*)([0-9]*)", tag_name)
+        g = m.groups()
```
Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: This code is still buggy. It ignores everything after the first group of digits like it did earlier. In other words, order of tags like `1a`, `1b` and `1c` is undefined. I'm not sure what's the best way to fix that.

Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: This version is actually even more buggy than the old one. Now tags that contain special characters like `foo-1` are not handled correctly.

**Comment by HelioGuilherme66** [comment]
Well, considering the initial objective, now the tags are listed in a better organized way (sorted). The test file is [Test_tags.robot](https://github.com/HelioGuilherme66/RIDE/raw/sort_tags_%231579/utest/resources/robotdata/view_all_tags/Tags_test.robot). Please say if you think that RF log file would benefit with this sorting variant (and should we put that under discussion in users group).

**Comment by peterlck** [comment]
Hi,

I would like to unsubscribe from the thread and I have clicked on the
Unsubscribe link numerous time but to no avail.
Can someone help me to unsubscribe ?

**Comment by HelioGuilherme66** [comment]
@peterlck You are getting this from GitHub not a User Group mailing list.
To stop receiving this, just click the Unwatch button in this project [RIDE](https://github.com/robotframework/RIDE/)

**Comment by HelioGuilherme66** [comment]
@yanne @pekkaklarck @mkorpela 
Finally the code is ready to be merged.
Algorithm is simplified and unitest was added.
Includes fix #1621

File: src/robotide/ui/tagdialogs.py [hunk 7]
```diff
@@ -100,6 +101,16 @@ def _make_bindings(self):
         self._tags_list.Bind(wx.EVT_LIST_ITEM_RIGHT_CLICK, self.OnRightClick)
         self._tags_list.Bind(wx.EVT_LIST_COL_CLICK, self.OnColClick)
 
+    def _tag_name_for_sort(self, tag_name):
+        mylist = [part if index % 2 == 0 else int(part) for index, part in
+                  enumerate(re.split(r'(\d+)', tag_name.lower()))]
```
Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: This is a good way to split the name into comparable tokens. I would rename `mylist` to `tokens`, though.

File: src/robotide/ui/tagdialogs.py [hunk 8]
```diff
@@ -100,6 +101,16 @@ def _make_bindings(self):
...
+        # Side effect from using regex, u'' will be removed from start or end
+        if mylist[0] == unicode(''):
+            del mylist[0]
+        if mylist[-1] == unicode(''):
+            del mylist[-1]
```
Comment (pekkaklarck) [inline src/robotide/ui/tagdialogs.py]: Empty strings shouldn't be removed. Keeping them enforces that the returned sort key always starts and ends with a string and is thus comparable. Sorting, for example, `('foo', 1)` and `(2, 'bar')` makes no sense (in Python 3 it even raises `TypeError`), but comparing `('foo', 1, '')` and `('', 2, 'bar')` is fine.

**Comment by pekkaklarck** [comment]
1. The latest version looks pretty good. See my line note above for one place that ought to be fixed simply by removing few lines of code.

2. It is not a good idea to have unrelated changes in a commit or PR. There aren't too many of them so I'm fine leaving them this time.

3. Commits in this PR should be squashed.

**Comment by HelioGuilherme66** [comment]
Squashed and final :)
Thanks!
